File: client/src/modules/nexotype/schemas/asset/biologic.schemas.ts
import { z } from 'zod';

// ==========================================
// Enums
// ==========================================

export const BiologicTypeEnum = z.enum(['Antibody', 'Enzyme']);

// ==========================================
// Options (for Select dropdowns)
// ==========================================

/** Biologic.biologic_type */
export const BIOLOGIC_TYPE_OPTIONS = [
  { label: 'Antibody', value: 'Antibody' },
  { label: 'Enzyme', value: 'Enzyme' },
] as const;

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const BiologicSchema = z.object({
  id: z.number(),
  uid: z.string().min(1).max(100),
  name: z.string().min(1).max(255),
  project_code: z.string().max(50).nullable().optional(),
  asset_type: z.string().max(50),
  sequence_aa: z.string().min(1),
  biologic_type: BiologicTypeEnum,
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateBiologicSchema = z.object({
  uid: z.string().min(1).max(100),
  name: z.string().min(1).max(255),
  project_code: z.string().max(50).nullable().optional(),
  sequence_aa: z.string().min(1),
  biologic_type: BiologicTypeEnum,
});

export const UpdateBiologicSchema = z.object({
  uid: z.string().min(1).max(100).optional(),
  name: z.string().min(1).max(255).optional(),
  project_code: z.string().max(50).nullable().optional(),
  sequence_aa: z.string().min(1).optional(),
  biologic_type: BiologicTypeEnum.optional(),
});

// ==========================================
// Types
// ==========================================

export type Biologic = z.infer<typeof BiologicSchema>;
export type CreateBiologic = z.infer<typeof CreateBiologicSchema>;
export type UpdateBiologic = z.infer<typeof UpdateBiologicSchema>;
export type BiologicType = z.infer<typeof BiologicTypeEnum>;

// ==========================================
// Response Types
// ==========================================

export type BiologicResponse = {
  success: boolean;
  data?: Biologic;
  error?: string;
};

export type BiologicsResponse = {
  success: boolean;
  data?: Biologic[];
  count?: number;
  error?: string;
};

File: server/apps/nexotype/schemas/asset/biologic_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class BiologicCreate(BaseModel):
    """Schema for creating a new biologic"""
    uid: str
    name: str
    project_code: str | None = None
    sequence_aa: str
    biologic_type: str


class BiologicUpdate(BaseModel):
    """Schema for updating a biologic"""
    uid: str = None
    name: str = None
    project_code: str | None = None
    sequence_aa: str = None
    biologic_type: str = None


# ==========================================
# Response Schemas
# ==========================================

class BiologicDetail(BaseModel):
    """Schema for biologic details"""
    id: int
    uid: str
    name: str
    project_code: str | None = None
    asset_type: str
    sequence_aa: str
    biologic_type: str
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class BiologicListResponse(BaseModel):
    """Response schema for listing biologics"""
    success: bool
    data: list[BiologicDetail] | None = None
    count: int = 0
    error: str | None = None


class BiologicResponse(BaseModel):
    """Response schema for single biologic operations"""
    success: bool
    data: BiologicDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None
